server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfileProjectsActionTest.java

```diff
@@ -0,0 +1,224 @@
+      .execute().assertJson(this.getClass(), "empty.json");
+  }
+
+  @Test
+  public void should_show_unselected() throws Exception {
+    project1 = newProject("ABCD", "Project One");
+    project2 = newProject("BCDE", "Project Two");
+    project3 = newProject("CDEF", "Project Three");
+    project4 = newProject("DEFA", "Project Four");
+    dbClient.componentDao().insert(session, project1, project2, project3, project4);
+
+    addBrowsePermissionToAnyone(session, project1, project2, project3, project4);
+
+    associateProjectsWithProfile(session, xooP1, project1, project2);
+
+    session.commit();
+
+    newRequest().setParam("key", xooP1.getKey()).setParam("selected", "deselected").execute().assertJson(this.getClass(), "deselected.json");
+  }
+
+  @Test
+  public void should_show_all() throws Exception {
+    project1 = newProject("ABCD", "Project One");
+    project2 = newProject("BCDE", "Project Two");
+    project3 = newProject("CDEF", "Project Three");
+    project4 = newProject("DEFA", "Project Four");
+    dbClient.componentDao().insert(session, project1, project2, project3, project4);
+
+    addBrowsePermissionToAnyone(session, project1, project2, project3, project4);
+
+    associateProjectsWithProfile(session, xooP1, project1, project2);
+    // project3 is associated with P2, must appear as not associated with xooP1
+    associateProjectsWithProfile(session, xooP2, project3);
+
+    session.commit();
+
+    newRequest().setParam("key", xooP1.getKey()).setParam("selected", "all").execute().assertJson(this.getClass(), "all.json");
+  }
+
+  @Test
+  public void should_filter_on_name() throws Exception {
+    project1 = newProject("ABCD", "Project One");
+    project2 = newProject("BCDE", "Project Two");
+    project3 = newProject("CDEF", "Project Three");
+    project4 = newProject("DEFA", "Project Four");
+    dbClient.componentDao().insert(session, project1, project2, project3, project4);
+
+    addBrowsePermissionToAnyone(session, project1, project2, project3, project4);
+
+    associateProjectsWithProfile(session, xooP1, project1, project2);
+
+    session.commit();
+
+    newRequest().setParam("key", xooP1.getKey()).setParam("selected", "all").setParam("query", "project t").execute().assertJson(this.getClass(), "all_filtered.json");
+  }
+
+  @Test(expected = NotFoundException.class)
+  public void should_fail_on_unexisting_profile() throws Exception {
+    newRequest().setParam("key", "unknown").setParam("selected", "all").execute();
+  }
+
+  private void createProfiles() {
+    xooP1 = QProfileTesting.newXooP1();
+    xooP2 = QProfileTesting.newXooP2();
+    dbClient.qualityProfileDao().insert(session, xooP1, xooP2);
+  }
+
+  private TestRequest newRequest() {
+    return wsTester.newGetRequest("api/qualityprofiles", "projects");
+  }
+
+  private ComponentDto newProject(String uuid, String name) {
+    return ComponentTesting.newProjectDto().setUuid(uuid).setName(name);
```

Comment:
It's better to use ComponentTesting.newProjectDto(uuid), otherwise the uuid of the project will be different than the projectUuid, and you have strange results if you create 2 projects with newProjectDto().setUuid(uuid)